Websocket support

rshriram opened the pull request:
This PR adds WebSocket support to Envoy. 
The main implementation is fairly complete. I have tested this code with a simple python based websocket client talking to echo.websocket.org (both 80 and 443) and it works :). I would appreciate any feedback on the code. 

Tests are TODO. Finding it difficult, as this requires part HTTP mocks and part TCP mocks. Pointers appreciated.

Need to test interaction of WebSockets in the presence of RDS (cc @kyessenov @GregHanson @ijsnellf)

cc @mattklein123 @htuch @PiotrSikora

rshriram commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -466,6 +485,56 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
+  // would require us to check for upgrades in encodeHeaders where the machinery
+  // gets lot more complicated. So, for the moment, when we have outstanding responses,
+  // we reject the upgrade request. From a browser point of view, Chrome/Mozilla disable
+  // pipelining, while IE11+ does not even support pipelining. So this limitation should
```
@PiotrSikora @mattklein123 any thoughts on this? If this is not going to be a problem, I would happily get rid of this counter business.

alyssawilk replied:
I *thought* the readDisable was because Envoy only handled one http request a a time, so didn't do true pipelining (at which point you'd be able to drop all of the response tracking).  New enough to the code base that I can't claim with certainty that it's true though  - Matt would know for sure :-)

mattklein123 replied:
Correct, we do not do real pipelining in the server since I did not want to deal with out of order responses for HTTP/1.1. We do allow the client to pipeline and handle that correctly. I don't think we will ever support full HTTP/1.1 pipelining in Envoy. At this point people can use HTTP/2 if they care. For HTTP/1.1 there will never be any pending responses. It will always be one request at a time until a response is sent.

mattklein123 commented on source/common/http/websocket/ws_handler_impl.h:
```diff
@@ -0,0 +1,93 @@
+ * N.B. This class implements Network::ReadFilter interfaces purely for sake of consistency
+ * with TcpProxy filter. WsHandlerImpl it is not used as a network filter in any way.
+ */
+class WsHandlerImpl : public Network::ReadFilter, Logger::Loggable<Logger::Id::websocket> {
```
Very quick comment: I would like to see if there is a way to share the code with tcp_proxy, either via a shared utility class or consuming tcp_proxy internally much like async_client consumes the router filter.

rshriram replied:
Yes.. The main issue is that while the TCP proxy closes downstream connections on error, the websocket class sends http errors to downstream. So, while the control flow is same, the error recovery path is different. But i will still try where possible.

rshriram replied:
That is a good point. My aim is to move most of ws_handler out, and reuse TCP. However, the crux of ws_handler is in initializeUpstreamConnection (where all this error handling is done in Websocket specific way - for example, no tcp backend, send http 503/404).

I think I had noted this elsewhere in the PR. If there is a way to latch onto a backend connection for a request (we would need this for sticky sessions anyway, if we plan to support it), then this entire WS implementation can reuse HTTP/1.1 code, and drop into tcp only when the websocket upgrade is complete.

alyssawilk commented on test/integration/integration_test.cc:
```diff
@@ -141,6 +141,97 @@ TEST_P(IntegrationTest, OverlyLongHeaders) {
+}
+
+#if 0
+  // HELP: Eventhough the initial part of this test is identical to the other websocket
```
Not sure if this helps any but last time I hit this I literally copy-pasted the old test and replaced my new one line by line until I found the difference.
Good luck either way!

rshriram commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -91,6 +93,9 @@ ConnectionManagerImpl::~ConnectionManagerImpl() {
+      if (isWebSocketConnection()) {
+        stats_.named_.downstream_cx_websocket_active_.dec();
+      }
       stats_.named_.downstream_cx_http1_active_.dec();
```
Yes, i thought about it and my rational was that WebSocket is strictly a subset of a HTTP/1.1 connection, and as such, http1 counters should be reflected as well. That said, I like your point about total connections to a machine.

dnoe commented on docs/configuration/http_conn_man/route_config/route.rst:
```diff
@@ -117,6 +118,24 @@ case_sensitive
+    true, Envoy will setup plain TCP proxying between the client and the upstream server. Hence,
+    an upstream server that rejects the WebSocket upgrade request is also responsible for closing
+    the associated connection. Until then, Envoy will continue to proxy data from the client to
+    the upstream server.
```
This seems a little strange to me, is there precedent for other websocket proxying working this way? If a WebSocket upstream server is rejecting the WebSocket upgrade request because of some overload/error condition will data pile up in Envoy?

rshriram replied:
If the upstream closes its connection after rejecting the upgrade, Envoy will flush the pipe and close downstream as well (standard tcp proxy semantics). Until upstream keeps the connection open, Envoy does standard tcp proxying - buffer bloat is subject to flow control settings.

dnoe replied:
OK, I think that sounds fine.

dnoe commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -470,6 +488,60 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
+      // there is an error establishing connection to upstream.
+      // TODO (rshriram) - can decoder_filters_ be empty? Is there a better way to get
+      // a handle over the stream decoder filter callback instead of ** ?
+      std::list<ActiveStreamDecoderFilterPtr>::iterator entry = decoder_filters_.begin();
```
Your call but I'd always use auto for iterator types.

rshriram replied:
I am anti auto due to scars from Golang. ;)
and this is a smallish type name. I don't mind changing this if you feel strongly otherwise.

dnoe replied:
The Google style guide that we inherit Envoy style from says that auto is "encouraged" for iterator types, so I'd prefer it.

One more thing to think about here is whether the iterator can be const. I think it probably can be because WsHandlerImpl takes a const pointer right now.

dnoe commented on source/common/http/websocket/ws_handler_impl.cc:
```diff
@@ -0,0 +1,145 @@
+namespace WebSocket {
+
+WsHandlerImpl::WsHandlerImpl(Http::HeaderMap& request_headers,
+                             const Router::RouteEntry* route_entry,
```
Can it be a const reference, or should it be a shared or unique_ptr instead if it needs to be nullable?

rshriram replied:
Const reference sounds like a good idea. Route entries are read only

rshriram commented on source/common/filter/tcp_proxy.cc:
```diff
@@ -140,6 +140,10 @@ void TcpProxy::readDisableUpstream(bool disable) {
 
 void TcpProxy::readDisableDownstream(bool disable) {
   read_callbacks_->connection().readDisable(disable);
+  if (!config_) {
```
My bad. I should have added a comment. I am reusing the tcp proxy code in ws_handler_impl, where I initialize the tcp_proxy class with a null config.

rshriram commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -470,6 +488,60 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
+  }
+
+  // Check for WebSocket upgrade request if the route is valid, and supports WebSockets.
+  if ((protocol == Protocol::Http11) && cached_route_.value()) {
```
I was checking if the route was a null pointer. In retrospect, I can reuse the same valid() call to check if it's a nullptr.

rshriram replied:
update. I need to use the value() to check if the snapped_route_config_->route actually returned a non-null route.

rshriram commented on source/common/http/conn_manager_impl.h:
```diff
@@ -39,6 +40,7 @@ namespace Http {
   COUNTER(downstream_cx_total)                                                                     \
   COUNTER(downstream_cx_ssl_total)                                                                 \
   COUNTER(downstream_cx_http1_total)                                                               \
+  COUNTER(downstream_cx_websocket_total)                                                           \
```
Done

dnoe commented on source/common/json/config_schemas.cc:
```diff
@@ -606,6 +606,7 @@ const std::string Json::Schema::ROUTE_ENTRY_CONFIGURATION_SCHEMA(R"EOF(
       "prefix_rewrite" : {"type" : "string"},
       "host_rewrite" : {"type" : "string"},
       "auto_host_rewrite" : {"type" : "boolean"},
+      "allow_websocket_upgrade" : {"type" : "boolean"},
```
If this is set we actually **mandate** websocket upgrade, right? Otherwise we fail the request. The config field name should reflect that if possible, maybe just call it "websocket" or "use_websockets"?

dnoe commented on source/common/router/config_impl.cc:
```diff
@@ -558,6 +564,8 @@ const VirtualHostImpl* RouteMatcher::findVirtualHost(const Http::HeaderMap& head
   }
 
+  // TODO (@rshriram) Match Origin header in WebSocket
+  // request with VHost, using wildcard match
```
Is this a TODO that should be completed before merging this PR?

dnoe commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -1039,6 +1094,9 @@ TEST_F(HttpConnectionManagerImplTest, MultipleFilters) {
 
   setupFilterChain(3, 2);
 
+  // Test route caching.
```
I'm not clear on what this move has to do with the websocket change?

rshriram replied:
In conn_manager_impl.cc, when checking for websocket upgrade compatibility, we are fetching the routes and checking for the "allow_websocket_upgrade" flag. Since websocket introduces a call to route() way before entering any route filter, this EXPECT_CALL has to be moved before the expects for filters. Otherwise, the code segfaults. :(

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -470,6 +488,62 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
+        connection_manager_.stats_.named_.downstream_cx_http1_active_.dec();
+      } else {
+        // TODO (rshriram) - this can only happen if someone has no router filter. Not sure if its
+        // valid in the config.
```
I haven't looked through the code yet, but no router filter should be required for websocket to work. If it is something is probably not right. I can look more later. Sorry I have a large review backlog.

rshriram commented on source/common/http/websocket/ws_handler_impl.cc:
```diff
@@ -0,0 +1,145 @@
+      stream_.encodeHeaders(std::move(headers), true);
+    }
+
+    read_callbacks_->connection().close(Network::ConnectionCloseType::FlushWrite);
```
Good question. Unfortunately, I just took this code verbatim out of common/filters/tcp_proxy.cc and I assumed that whatever code was there did all the due diligence. 

The only reason I had to duplicate the code was because of the need for a different error handling compared to standard TCP semantics (close connections).

dnoe replied:
This feels like a lot of code taken largely verbatim from TCPProxy, along with the other spots mentioned above. It would be really nice to avoid duplicating implementation. I suspect that the difference in error handling could be implemented by adding additional overrides to the interface for TCPProxy at the points where you need different error handling behavior and calling the virtual function from TCPProxy base at the error point.

rshriram commented on source/common/http/websocket/ws_handler_impl.cc:
```diff
@@ -0,0 +1,145 @@
+    // Once we see a 101 switching protocols, we could remove the connection from
+    // the connection pool. The current approach is a stop gap solution, where
+    // we put the onus on the user to tell us if a route (and corresponding upstream)
+    // is supposed to allow websocket upgrades or not.
```
Yes, user is the person configuring Envoy. And I have the same concern here as well.

rshriram commented on source/common/http/websocket/ws_handler_impl.h:
```diff
@@ -0,0 +1,56 @@
+ * the underlying WebSocket protocol.
+ *
+ * N.B. This class implements Network::ReadFilter interfaces purely for sake of consistency
+ * with TcpProxy filter. WsHandlerImpl it is not used as a network filter in any way.
```
I think this entire comment should go. Its outdated and no longer applicable. Sorry about that.

rshriram commented on source/common/http/conn_manager_impl.h:
```diff
@@ -537,6 +545,8 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
   Tracing::HttpTracer& tracer_;
   Runtime::Loader& runtime_;
   const LocalInfo::LocalInfo& local_info_;
+  Upstream::ClusterManager& cluster_manager_;
```
I can double check. The TCP proxy code seemed to do the same.

rshriram commented on source/common/http/websocket/ws_handler_impl.cc:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  if (event & Network::ConnectionEvent::RemoteClose) {
+    if (connect_timeout_timer_) {
```
"Can the event bitmask include events Connected and RemoteClose? Since the Connected handling is below we won't have cleared out the timer so we'll send BadGateway. Is that intended?
"

I don't think the bitmask can include connected and remoteclose. Atleast, from looking at the TCP proxy code and the http conn man code, I don't see a precedent for this. And this code repeats the pattern found in
https://github.com/lyft/envoy/blob/master/source/common/filter/tcp_proxy.cc#L266 (TCP_PROXY)

dnoe commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -470,6 +488,62 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
+      std::list<ActiveStreamDecoderFilterPtr>::iterator entry = decoder_filters_.begin();
+      if (entry != decoder_filters_.end()) {
+        connection_manager_.ws_connection_ =
+            std::unique_ptr<WebSocket::WsHandlerImpl>(new WebSocket::WsHandlerImpl(
```
Can you avoid some of this kludgeyness and not use an existing filter to do encodeHeaders() by doing something like this:

1) Make a new interface, something like WebsocketHandlerCallbacks, with a method like sendHeadersOnlyResponse() - this is nice because you can mock it for tests.

2) Have your implementation of WebsocketHandlerCallbacks hold the WsHandlerImpl.

3) Connection Manager can instantiate WebsocketHandlerCallbacks impl, get headers with the sendHeadersOnlyResponse() method, call encodeHeaders, and then forward to the parent active stream.

dnoe replied:
(the idea here is to avoid too much Websocket implementation leaking into ConnectionManager and also eliminate the dependency on having an existing route filter, which should not be a requirement)

rshriram replied:
I just need a handle to encodeHeaders from the WebSocket Impl. May be make ActiveStream implement this callback interface and pass a reference to the WsHandlerCallbacks to WsHandlerImpl so that I don't have to do this kludgey conversion.
I cannot instantiate a WsHandlerImpl inside the callback, without having to pass the same parameters to the callback constructor. The reason is that when making upstream connection, WsHandler needs to pass on the request headers, and the route entry (to perform prefix/host rewrites.

rshriram commented:
@mattklein123 @dnoe I think i have addressed all of your comments (with the exception of one or two, for which I have provided inline responses). PTAL

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -470,6 +487,47 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
       connection_manager_.config_, *snapped_route_config_, connection_manager_.random_generator_,
       connection_manager_.runtime_, connection_manager_.local_info_);
 
+  if (!cached_route_.valid()) {
```
I don't think it's possible for it to be valid here? Can you switch this to an ASSERT followed by a direct set? Also, I think that means that elsewhere where route is fetched you can remove the guard and just return the value.

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -470,6 +487,47 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
+  // Check for WebSocket upgrade request if the route exists, and supports WebSockets.
+  if ((protocol == Protocol::Http11) && cached_route_.value()) {
+    const Router::RouteEntry* route_entry = cached_route_.value()->routeEntry();
+    // Websocket route entries cannot be redirects.
```
Can we check this in config_impl and then just fail there? Then these can just be asserts.

rshriram replied:
I already check these in config_impl. I think this comment is misleading.

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -470,6 +487,47 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
+    if (websocket_requested && websocket_required) {
+      ENVOY_STREAM_LOG(debug, "found websocket connection. (end_stream={}):", *this, end_stream);
+
+      connection_manager_.ws_connection_ =
```
nit: just do .reset(new ...) here. a bit less verbose.

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -470,6 +487,47 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
+      connection_manager_.stats_.named_.downstream_cx_websocket_active_.inc();
+      connection_manager_.stats_.named_.downstream_cx_websocket_total_.inc();
+      connection_manager_.stats_.named_.downstream_cx_http1_active_.dec();
+      return;
```
nit: returns can be lifted from both inner ifs out to around line 529.

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -470,6 +487,47 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
+  }
+
+  // Check for WebSocket upgrade request if the route exists, and supports WebSockets.
+  if ((protocol == Protocol::Http11) && cached_route_.value()) {
```
For someone that wants to just proxy web sockets, there is actually no reason that they need to install the router filter (or any filters at all). However, IIRC, if someone does not install the router filter, and there is no websocket match here, there will be no response at all. IMO, if there are no filters and we attempt to start filter iteration, the connection manager should return 404. Can you add a TODO to that effect?

rshriram replied:
Yes and a prior version of the PR actually sent a 404 if no filters were configured. But then, @dnoe pointed to docs that stated that route filter must be configured for HTTP conn man, and given that websocket is under http conn man, I removed it. I can add the check for lack of route filter for websockets but if you want it to be a http conn man wide feature, its probably done in a separate PR.

mattklein123 replied:
I would just add a TODO. We can fix in an independent change.

mattklein123 commented on source/common/http/websocket/ws_handler_impl.cc:
```diff
@@ -0,0 +1,143 @@
+    : Filter::TcpProxy(nullptr, cluster_manager), request_headers_(request_headers),
+      route_entry_(route_entry), ws_callbacks_(callbacks) {}
+
+WsHandlerImpl::~WsHandlerImpl() {}
```
nit: del since emtpy.

mattklein123 commented on source/common/http/websocket/ws_handler_impl.cc:
```diff
@@ -0,0 +1,143 @@
+    ENVOY_CONN_LOG(debug, "creating connection to upstream cluster {}",
+                   read_callbacks_->connection(), cluster_name);
+  } else {
+    HeaderMapImpl headers{{Headers::get().Status, std::to_string(enumToInt(Http::Code::NotFound))}};
```
I'm still not thrilled with how much code is copied here. Is it possible to refactor slightly so that the same init code runs, but if there is a failure during init you call an onInitFailure() virtual function that can do different stuff? I think that should work and be very clean and copy much less code.

rshriram commented on source/common/filter/tcp_proxy.cc:
```diff
@@ -140,6 +140,11 @@ void TcpProxy::readDisableUpstream(bool disable) {
 
 void TcpProxy::readDisableDownstream(bool disable) {
   read_callbacks_->connection().readDisable(disable);
+  // The WsHandlerImpl class uses TCP Proxy code with a null config.
```
The only difference between the TCP and HTTP downstream connection stats were the flow control stats. Others like 
```c++
  COUNTER(downstream_cx_rx_bytes_total)                                                            \
  GAUGE  (downstream_cx_rx_bytes_buffered)                                                         \
  COUNTER(downstream_cx_tx_bytes_total)                                                            \
  GAUGE  (downstream_cx_tx_bytes_buffered)                                                         \
  COUNTER(downstream_cx_total)                                                                     \
```
are already being recorded by HTTP. I was under the assumption that when flow control is added to HTTP, these missing stats would naturally fall in.

mattklein123 replied:
Hmm OK yes I think you are right. Can you add TODO for @alyssawilk here so we make sure this works? (Or maybe it already does in the outstanding PR).

rshriram commented:
@mattklein123 @dnoe addressed comments.

dnoe commented:
Will take a look today, thanks.

dnoe commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -487,31 +487,32 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
+        connection_manager_.ws_connection_->initializeReadFilterCallbacks(
+            *connection_manager_.read_callbacks_);
+        if (connection_manager_.ws_connection_->onNewConnection() ==
+            Network::FilterStatus::Continue) {
```
Is there an error case here?

rshriram replied:
the onNewConnection will send the error response back to caller via the WsHandlerCallbacks.

mattklein123 commented on source/common/filter/tcp_proxy.h:
```diff
@@ -133,6 +138,46 @@ class TcpProxy : public Network::ReadFilter, Logger::Loggable<Logger::Id::filter
+  struct UpstreamHelperCallbacks {
+    virtual ~UpstreamHelperCallbacks() {}
+
+    virtual const std::string& getUpstreamCluster() PURE;
```
Is there a reason that we need to have a new object here that we allocate? Can these just be normal virtual functions that get overriden by the WsHandlerImpl derived class? That would save the extra allocation which AFAICT isn't needed?

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -91,7 +93,11 @@ ConnectionManagerImpl::~ConnectionManagerImpl() {
     } else {
-      stats_.named_.downstream_cx_http1_active_.dec();
+      if (isWebSocketConnection()) {
+        stats_.named_.downstream_cx_websocket_active_.dec();
```
Optimally, you would verify this stat dance in your integration tests. Can you make sure that the gauges are correct at various points?

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -470,6 +487,47 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
+  // should return 404. The current returns no response if there is no router filter.
+  if ((protocol == Protocol::Http11) && cached_route_.value()) {
+    const Router::RouteEntry* route_entry = cached_route_.value()->routeEntry();
+    bool websocket_required = (route_entry != nullptr) && route_entry->useWebSocket();
```
const for websocket_required and websocket_requested

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -470,6 +487,47 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
+
+        connection_manager_.ws_connection_.reset(new WebSocket::WsHandlerImpl(
+            *request_headers_, *route_entry, *this, connection_manager_.cluster_manager_));
+        connection_manager_.ws_connection_->initializeReadFilterCallbacks(
```
nit: Can this be done as part of the WsHandlerImpl constructor?

mattklein123 commented on source/common/http/websocket/ws_handler_impl.h:
```diff
@@ -0,0 +1,70 @@
+ */
+class WsHandlerImpl : public Filter::TcpProxy {
+public:
+  WsHandlerImpl(Http::HeaderMap& request_headers, const Router::RouteEntry& route_entry,
```
nit: Since you are in Http namespace, you shouldn't need `Http::` in all/most places in this file and cc file. Can you remove?

mattklein123 commented on source/common/http/websocket/ws_handler_impl.h:
```diff
@@ -0,0 +1,70 @@
+  }
+
+protected:
+  struct WsUpstreamHelperCallbacks : public UpstreamHelperCallbacks {
```
per other comment I think these can just be virtual overrides.

mattklein123 commented on source/common/http/websocket/ws_handler_impl.h:
```diff
@@ -0,0 +1,70 @@
+private:
+  struct NullHttpConnectionCallbacks : public Http::ConnectionCallbacks {
+    // Http::ConnectionCallbacks
+    void onGoAway() override{};
```
nit: remove ';' at end of line

mattklein123 commented on source/common/http/websocket/ws_handler_impl.h:
```diff
@@ -0,0 +1,70 @@
+public:
+  WsHandlerImpl(Http::HeaderMap& request_headers, const Router::RouteEntry& route_entry,
+                WsHandlerCallbacks& callbacks, Upstream::ClusterManager& cluster_manager);
+  ~WsHandlerImpl(){};
```
nit: del, not needed.

mattklein123 commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -502,6 +531,80 @@ TEST_F(HttpConnectionManagerImplTest, NoPath) {
+  EXPECT_EQ(1U, stats_.named_.downstream_cx_websocket_active_.value());
+  EXPECT_EQ(1U, stats_.named_.downstream_cx_websocket_total_.value());
+  EXPECT_EQ(0U, stats_.named_.downstream_cx_http1_active_.value());
+}
```
if you reset() conn_manager_ here can you then verify the gauges are zero? Same below?

rshriram replied:
It isn't working. Getting assertion error from ActiveStream in tsan and asan.

mattklein123 replied:
I don't know what the problem is off the top of my head. As long as you have coverage in the integration tests which it looks like you added, that's fine.

mattklein123 commented on test/integration/server.h:
```diff
@@ -199,6 +199,10 @@ class IntegrationTestServer : Logger::Loggable<Logger::Id::testing>,
   }
 
+  Stats::GaugeSharedPtr gauge(const std::string& name) {
+    return TestUtility::findGauge(*stat_store_, name);
```
The same comment applies here as to counter() above. Can you add a similar comment or just have a comment that says something to the effect of "See counter() above for why this method exists."
